flax/linen/module.py

```diff
@@ -221,25 +220,31 @@ def __init_subclass__(cls):
     cls.scope = None
 
   @classmethod
-  def _add_parent_and_name_attrs(cls):
-    """Add final optional dataclass attributes: `parent` and `name`."""
-    annotations = cls.__dict__.get('__annotations__', {})
+  def _customized_dataclass_transform(cls):
+    """Handle final optional dataclass attributes: `parent` and `name`."""
+    annotations = dict(cls.__dict__.get('__annotations__', {}))
     if 'parent' in annotations or 'name' in annotations:
       raise ValueError(
           f'properties `parent` and `name` are reserved: {annotations}')
     # Add `parent` and `name` default fields at end.
-    new_annotations = {}
-    new_annotations.update(annotations)
-    if 'parent' in getattr(cls, '__dataclass_fields__', {}):
+    # We temporarily modify base class __dataclass_fields__ to force desired
+    # argument behavior and ordering from dataclass class-transform.
+    parent_dataclass_fields = dict(getattr(cls, '__dataclass_fields__', {}))
+    if 'parent' in parent_dataclass_fields:
       cls.__dataclass_fields__.pop('parent')
-    new_annotations['parent'] = Union[Type["Module"], Type["Scope"],
-                                      Type["_Sentinel"], None]
-    cls.parent = dataclasses.field(repr=False, default=_unspecified_parent)
-    if 'name' in getattr(cls, '__dataclass_fields__', {}):
+    if 'name' in parent_dataclass_fields:
       cls.__dataclass_fields__.pop('name')
-    new_annotations['name'] = str
-    cls.__annotations__ = new_annotations
+    annotations['parent'] = Union[Type["Module"], Type["Scope"],
+                                  Type["_Sentinel"], None]
+    cls.parent = dataclasses.field(repr=False, default=_unspecified_parent)
+    annotations['name'] = str
     cls.name = None  # default value of name is None.
+    cls.__annotations__ = annotations
+    # Now apply dataclass transform (which operates in-place).
+    dataclasses.dataclass(cls)
+    # Restore original base class __dataclass_fields__.
+    if dataclasses.is_dataclass(cls.__bases__[0]):
+     cls.__bases__[0].__dataclass_fields__ = parent_dataclass_fields
```

Comment:
Hmm are you sure this is correct? IIUC then `parent_dataclass_fields` contains dataclass fields of `cls` and its base class. Now you will set the dataclass fields of the base class to this, which means the base class seems to inherit the fields from the child class. That doesn't seem like the right behavior to me, or am I missing something?

Reply:
When parent_dataclass_fields is grabbed and copied by `dict(getattr(cls, '__dataclass_fields__', {}))` we _know_ there is no `__dataclass_fields__` on the child, because the `@dataclass` transform hasn't been run on this class yet.  So the only fields that can show up come from direct inheritance.
(Note there's the wrinkle of trying to deal w. multiple inheritance in this code, but that is always a buggy nightmare in python and we should just explicitly forbid multiple inheritance - dataclasses themselves only barely work in a useable way w. multiple inheritance.)